File: API/_Services/Services/PositionService.cs
using API._Repositories;
using API._Services.Interfaces;
using API.Dtos;
using API.Models;
using AutoMapper;
using AutoMapper.QueryableExtensions;
using LinqKit;
using Microsoft.EntityFrameworkCore;
using SD3_API.Helpers.Utilities;

namespace API._Services.Services
{
    public class PositionService : IPositionService
    {
        private readonly IRepositoryAccessor _repository;
        private readonly IMapper _mapper;
        private readonly MapperConfiguration _mapConfiguration;

        public PositionService(
            IRepositoryAccessor repository, 
            IMapper mapper, 
            MapperConfiguration mapConfiguration)
        {
            _repository = repository;
            _mapper = mapper;
            _mapConfiguration = mapConfiguration;
        }

        public async Task<bool> Create(PositionDto dataDto)
        {
            dataDto.Status = true;
            dataDto.Create_Time = DateTime.Now;
            var data = _mapper.Map<Position>(dataDto);
            _repository.Position.Add(data);
            return await _repository.Save();
        }

        public async Task<bool> Delete(PositionDto dataDto)
        {
            if(_repository.Member.FindAll(x => x.Position_Id == dataDto.Id).Any())
                return false;
            
            dataDto.Status = false;
            return await Update(dataDto);
        }

        public async Task<PaginationUtility<PositionDto>> GetDataPagination(PaginationParam pagination, string text, bool isPaging = true)
        {
            var pred = PredicateBuilder.New<Position>(x => x.Status.Value);
            if(!string.IsNullOrWhiteSpace(text))
            {
                text = text.ToLower();
                pred.And(x => x.Name.ToLower().Contains(text));
            }

            var data = _repository.Position.FindAll(pred).ProjectTo<PositionDto>(_mapConfiguration).AsNoTracking();
            return await PaginationUtility<PositionDto>.CreateAsync(data, pagination.PageNumber, pagination.PageSize, isPaging);
        }

        public async Task<List<KeyValuePair<int, string>>> GetListPosition()
        {
            var data = await _repository.Position.FindAll(x => x.Status.Value).AsNoTracking()
                .Select(x => new KeyValuePair<int, string>(
                    x.Id,
                    x.Name
                )).ToListAsync();
            return data;
        }

        public async Task<bool> Update(PositionDto dataDto)
        {
            dataDto.Update_Time = DateTime.Now;
            var data = _mapper.Map<Position>(dataDto);
            _repository.Position.Update(data);
            return await _repository.Save();
        }
    }
}

File: CMS/src/app/_core/_services/position.service.ts
import { Injectable } from '@angular/core';
import { environment } from '@env/environment';
import { HttpClient } from '@angular/common/http';
import { Position } from '@models/position';
import { PaginationResult, PaginationParam } from '@utilities/pagination-utility';
import { KeyValuePair } from '@utilities/key-value-pair';

@Injectable({
  providedIn: 'root'
})
export class PositionService {
  baseUrl = environment.apiUrl;
  constructor(
    private http: HttpClient
  ) { }

  create(positon: Position) {
    return this.http.post<boolean>(`${this.baseUrl}Position/Create`, positon);
  }

  update(positon: Position) {
    return this.http.put<boolean>(`${this.baseUrl}Position/Update`, positon);
  }

  delete(positon: Position) {
    return this.http.put<boolean>(`${this.baseUrl}Position/Delete`, positon);
  }

  getData(pagination: PaginationParam, text: string) {
    return this.http.get<PaginationResult<Position>>(`${this.baseUrl}Position/GetDataPagination`, { params: { ...pagination, text } });
  }

  getListPosition() {
    return this.http.get<KeyValuePair[]>(`${this.baseUrl}Position/GetListPosition`);
  }
}
